### PR TITLE
Honor repo default branch

### LINKED ISSUE #133: Error without "Master"
Hello Guys,
i have some Puppet-Modules they don't have a "Master" branch and when i try to run Modulesync i get an error, see below. May Be related to #121 #76.

```bash
bash-3.2$ bundler exec msync update --noop
Syncing funkmdeclient
Cloning repository fresh
Cloning from ssh://git@git.xxx.com:2222/pupng/funkmdeclient.git
Error while updating funkmdeclient
bundler: failed to load command: msync (/usr/local/lib/ruby/gems/2.3.0/bin/msync)
Git::GitExecuteError: git '--git-dir=/Users/steinbrueckri/Projects/msync/modules/funkmdeclient/.git' '--work-tree=/Users/steinbrueckri/Projects/msync/modules/funkmdeclient' checkout 'origin/master'  2>&1:error: pathspec 'origin/master' did not match any file(s) known to git.
  /usr/local/lib/ruby/gems/2.3.0/gems/git-1.3.0/lib/git/lib.rb:937:in `command'
  /usr/local/lib/ruby/gems/2.3.0/gems/git-1.3.0/lib/git/lib.rb:644:in `checkout'
  /usr/local/lib/ruby/gems/2.3.0/gems/git-1.3.0/lib/git/base.rb:306:in `checkout'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync/git.rb:32:in `switch_branch'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync/git.rb:48:in `pull'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync.rb:97:in `manage_module'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync.rb:134:in `block in update'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync.rb:132:in `each'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync.rb:132:in `update'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/lib/modulesync/cli.rb:115:in `update'
  /usr/local/lib/ruby/gems/2.3.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
  /usr/local/lib/ruby/gems/2.3.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
  /usr/local/lib/ruby/gems/2.3.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
  /usr/local/lib/ruby/gems/2.3.0/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
  /usr/local/lib/ruby/gems/2.3.0/bundler/gems/modulesync-e53079030501/bin/msync:8:in `<top (required)>'
  /usr/local/lib/ruby/gems/2.3.0/bin/msync:23:in `load'
  /usr/local/lib/ruby/gems/2.3.0/bin/msync:23:in `<top (required)>'
bash-3.2$ cd /Users/steinbrueckri/Projects/msync/modules/funkmdeclient/
bash-3.2$ git branch
* release/2018.0.00.00
```

### PR DESCRIPTION
**Pull request by asedge**
Fixes #133 #76.

I had to disable `Metrics/ModuleLength` because it was over by 11 lines but `Modulesync::Git` could use some refactoring I think.  I wasn't sure how far I should go with it so I figured I would file the PR and see where the discussion leads.

### CODE REVIEW

File: lib/modulesync/git.rb [hunk 1]
```diff
@@ -18,7 +18,17 @@ def self.remote_branch_differ?(repo, local_branch, remote_branch)
         repo.diff("#{local_branch}..origin/#{remote_branch}").any?
     end
 
+    def self.default_branch(repo)
+      symbolic_ref = repo.branches.find { |b| b.full =~ %r{remotes/origin/HEAD} }
```
Comment (ekohl) [inline lib/modulesync/git.rb]: This assumes you use origin as a remote. I always use ```upstream``` as a remote so I'd appreciate if that was overridable somehow but I do see that assumption is already present in a few places.

Comment (asedge) [reply lib/modulesync/git.rb]: You're right, it is hard coded in a number of places.  Shall we handle it in this PR or a future one?

Comment (ekohl) [reply lib/modulesync/git.rb]: Since it's pre-existing code I'd handle it in a separate PR.

File: lib/modulesync/git.rb [hunk 2]
```diff
@@ -100,6 +110,7 @@ def self.update(name, files, options)
       module_root = "#{options[:project_root]}/#{name}"
       message = options[:message]
       repo = ::Git.open(module_root)
+      options[:branch] ||= (repo.current_branch || 'master')
```
Comment (ekohl) [inline lib/modulesync/git.rb]: I dislike changing options this way and think a local variable is cleaner. Same below.

Comment (asedge) [reply lib/modulesync/git.rb]: You're right, that's crap.  I started to change it to a local variable and then put it back to this for some reason.  I know that I had some questions about why the branch is being checked out again here and in `update_noop` after `switch_branch` is called in `pull`.  Looking over the code in `ModuleSync::update` again I suppose `ModuleSync::Git::update_noop` could be called in offline mode so you may need to check out a branch then.

File: lib/modulesync/cli.rb [hunk 3]
```diff
@@ -53,8 +53,8 @@ class Base < Thor
                    :desc => 'A regular expression to skip repositories.'
       class_option :branch,
                    :aliases => '-b',
-                   :desc => 'Branch name to make the changes in. Defaults to master.',
-                   :default => CLI.defaults[:branch] || 'master'
+                   :desc => 'Branch name to make the changes in. Defaults to the default branch of the upstream repository.',
```
Comment (ekohl) [inline lib/modulesync/cli.rb]: And falls back to master, right?

File: lib/modulesync/git.rb [hunk 4]
```diff
@@ -95,12 +105,18 @@ def self.tag(repo, version, tag_pattern)
       repo.push('origin', tag)
     end
 
+    def self.checkout_branch(repo, branch)
```
Comment (ekohl) [inline lib/modulesync/git.rb]: The name is quite similar to ```switch_branch```. Maybe this should be a private method to clearly indicate it's a helper?

**Comment by asedge** [comment]
@ekohl I have made those changed you requested.

**Comment by asedge** [comment]
@ekohl It's been a couple weeks.  Either it's fine or nobody cares.  I suppose both could be true. :smile:

**Comment by asedge** [comment]
@ekohl I'm fine if you want to get others to review this.  Just want to make sure it's not forgotten.  Thanks!

**Comment by ekohl** [comment]
Apologies, it's been a very busy couple of weeks.

**Comment by ekohl** [comment]
Thanks a lot!

**Comment by asedge** [comment]
I can appreciate that, thanks for the review/merge and for (I assume) volunteering your time!

**Pull request merged**